src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -433,22 +429,40 @@ async def run_setup_py(req: RunSetupPyRequest, setuptools: Setuptools) -> RunSet
             interpreter_constraints=req.interpreter_constraints,
         ),
     )
+
+    # We prefix the entire chroot, and run with this prefix as the cwd, so that we can capture any
+    # changes setup made within it (e.g., when running 'develop') without also capturing other
+    # artifacts of the pex process invocation.
+    chroot_prefix = "chroot"
+
     # The setuptools dist dir, created by it under the chroot (not to be confused with
     # pants's own dist dir, at the buildroot).
-    dist_dir = "dist/"
+    dist_dir = "dist"
+
+    prefixed_chroot = await Get(Digest, AddPrefix(req.chroot.digest, chroot_prefix))
+
+    # setup.py basically always expects to be run with the cwd as its own directory
+    # (e.g., paths in it are relative to that directory). This is true of the setup.py
+    # we generate and is overwhelmingly likely to be true for existing setup.py files,
+    # as there is no robust way to run them otherwise.
+    setup_script_reldir, setup_script_name = os.path.split(req.chroot.setup_script)
+    working_directory = os.path.join(chroot_prefix, setup_script_reldir)
+
     result = await Get(
         ProcessResult,
         VenvPexProcess(
             setuptools_pex,
-            argv=(req.chroot.setup_script, *req.args),
-            input_digest=req.chroot.digest,
+            argv=(setup_script_name, *req.args),
+            input_digest=prefixed_chroot,
+            working_directory=working_directory,
             # setuptools commands that create dists write them to the distdir.
             # TODO: Could there be other useful files to capture?
-            output_directories=(dist_dir,),
+            output_directories=(dist_dir,),  # Relative to the working_directory.
             description=f"Run setuptools for {req.exported_target.target.address}",
             level=LogLevel.DEBUG,
         ),
     )
+    # Note that output_digest paths are relative to the working_directory.
```

Comment:
This is different from #12359 - the error there was that I tried to strip "chroot/dist" when it should have just been "dist" (turns out output files are not just specified relative to the working dir, but the output digest is as well).